internal/worker/sshsession/worker.go

```diff
@@ -0,0 +1,364 @@
+	// and using this file allows us to not interfere with existing setups. The user may
+	// still add another file to the "AuthorizedKeysFile" directive in sshd_config.
+	//
+	// TODO(ale8k): At somepoint, update juju documentation to explain should they wish
+	// to ssh directly to a machine, this is the way to do so.
+	AuthorizedKeysFile = "authorized_keys2"
+)
+
+// FacadeClient holds the facade methods for the SSH session worker.
+type FacadeClient interface {
+	// WatchSSHConnRequest returns a watcher that will return doc ids for
+	// incoming SSH connection requests. These doc ids are to retrieve connect
+	// params from GetSSHConnRequest(docID string) (SSHConnRequest, error).
+	WatchSSHConnRequest(machineId string) (watcher.StringsWatcher, error)
+
+	// GetSSHConnRequest returns the SSH connection request for the given docID.
+	GetSSHConnRequest(docID string) (DummyParams, error)
+}
+
+// WorkerConfig encapsulates the configuration options for
+// instantiating a new lease ssh session worker.
+type WorkerConfig struct {
+	Logger           Logger
+	FacadeClient     FacadeClient
+	Agent            agent.Agent
+	ConnectionGetter ConnectionGetter
+}
+
+// Validate checks whether the worker configuration settings are valid.
+func (cfg WorkerConfig) Validate() error {
+	if cfg.Logger == nil {
+		return errors.NotValidf("nil Logger")
+	}
+	if cfg.FacadeClient == nil {
+		return errors.NotValidf("nil FacadeClient")
+	}
+	if cfg.Agent == nil {
+		return errors.NotValidf("nil Agent")
+	}
+	if cfg.ConnectionGetter == nil {
+		return errors.NotValidf("nil ConnectionGetter")
+	}
+
+	return nil
+}
+
+// sshSessionWorker is a worker that enables SSH connections to a machine.
+// It watches for incoming SSH connections to the controller's SSH server over
+// a facade, and when notified initiates a connection to the controller's SSH server.
+// When the controller's SSH server receives this connection, it initiates its own
+// SSH connection down the tunnel and the sshSessionWorker pipes it to the SSHD
+// on the machine.
+type sshSessionWorker struct {
+	tomb tomb.Tomb
+
+	logger           Logger
+	facadeClient     FacadeClient
+	agent            agent.Agent
+	connectionGetter ConnectionGetter
+}
+
+// NewWorker returns an SSH session worker.
+func NewWorker(cfg WorkerConfig) (worker.Worker, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &sshSessionWorker{
+		logger:           cfg.Logger,
+		facadeClient:     cfg.FacadeClient,
+		agent:            cfg.Agent,
+		connectionGetter: cfg.ConnectionGetter,
+	}
+
+	w.tomb.Go(w.loop)
+	return w, nil
+}
+
+// loop starts the workers main loop. It watches for incoming SSH connection requests
+// and handles them by initiating a reverse SSH connection to the controller and piping
+// it to the local sshd of the machine agent's machine.
+func (w *sshSessionWorker) loop() error {
+	machineId := w.agent.CurrentConfig().Tag().Id()
```

Comment:
This leaks implementation detail - define the machine id as an attribute on the worker and set if from the manifold

Reply:
Done